frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -980,19 +982,74 @@ TypeInference::checkExternConstructor(const IR::Node* errorPosition,
         errorPosition->srcInfo, new IR::Vector<IR::Type>(), rettype, args);
-    auto tvs = unify(errorPosition, mt, callType,
+    auto tvs = unify(errorPosition, methodType, callType,
                      "Constructor invocation %1% does not match constructor declaration %2%",
                      { callType, constructor });
     BUG_CHECK(tvs != nullptr || ::errorCount(), "Unification failed with no error");
     if (tvs == nullptr)
-        return nullptr;
+        return none;
+
+    LOG2("Constructor type before specialization " << methodType << " with " << tvs);
+    TypeVariableSubstitutionVisitor substVisitor(tvs);
+    substVisitor.setCalledBy(this);
+    auto specMethodType = methodType->apply(substVisitor);
+    LOG2("Constructor type after specialization " << specMethodType);
+    learn(specMethodType, this);
+    auto canon = getType(specMethodType);
+    if (canon == nullptr)
+        return none;
+
+    auto functionType = specMethodType->to<IR::Type_MethodBase>();
+    BUG_CHECK(functionType != nullptr, "Method type is %1%", specMethodType);
+    if (!functionType->is<IR::Type_Method>())
+        BUG("Unexpected type for function %1%", functionType);
 
     ConstantTypeSubstitution cts(tvs, refMap, typeMap, this);
-    auto newArgs = cts.convert(arguments);
-    return newArgs;
+    // Arguments may need to be cast, e.g., list expression to a
+    // header type.
+    auto paramIt = functionType->parameters->begin();
+    auto newArgs = new IR::Vector<IR::Argument>();
+    bool changed = false;
+    for (auto arg : *arguments) {
+        cstring argName = arg->name.name;
+        bool named = !argName.isNullOrEmpty();
+        const IR::Parameter* param;
+
+        if (named) {
+            param = functionType->parameters->getParameter(argName);
+        } else {
+            param = *paramIt;
+        }
+
+        auto newExpr = arg->expression;
+        if (param->direction == IR::Direction::In || param->direction == IR::Direction::None) {
+            // This is like an assignment; may make additional conversions.
+            newExpr = assignment(arg, param->type, arg->expression);
+        } else {
+            // Insert casts for 'int' values.
+            newExpr = cts.convert(newExpr)->to<IR::Expression>();
+       }
+        if (::errorCount() > 0)
+            return none;
+        if (newExpr != arg->expression) {
+            LOG2("Changing method argument to " << newExpr);
+            changed = true;
+            newArgs->push_back(new IR::Argument(arg->srcInfo, arg->name, newExpr));
+        } else {
+            newArgs->push_back(arg);
+        }
+        if (!named)
+            ++paramIt;
+    }
+    if (changed)
+        arguments = newArgs;
+    auto objectType = new IR::Type_Extern(ext->srcInfo, ext->name, methodType->typeParameters,
+                                          ext->methods);
+    return std::pair<const IR::Type*, const IR::Vector<IR::Argument>*>(objectType, arguments);
```

Comment:
Since p4c is now compiling as C++17 you can just do:
`return {objectType, arguments};`
here.